enhancements/ingress/set_eip_nlb_ingress.md

```diff
@@ -0,0 +1,1319 @@
+---
+title: set_eip_nlb_ingress
+authors:
+  - miheer
+reviewers:
+  - Miciah
+  - frobware
+  - gspence
+  - candace
+approvers:
+  - Miciah
+  - frobware
+  - gspence
+api-approvers:
+  - joel
+  - deads
+creation-date: 2024-03-06
+last-updated: 2024-03-22
+tracking-link:
+  - "https://issues.redhat.com/browse/NE-1274"
+see-also:
+  - ""
+replaces:
+  - ""
+superseded-by:
+  - ""
+---
+
+# Set AWS EIP For NLB Ingress Controller
+
+## Summary
+
+This enhancement allows user to set AWS EIP for the NLB default or custom ingress controller.
+This is a feature request to enhance the IngressController API to be able to support static IPs during
+- Install time
+- Custom NLB ingress controller creation 
+- Reconfiguration of the custom and default router to NLB with EIP allocation.
+
+## Motivation
+
+### User Stories
+
+- As a cluster administrator, I want to provision EIPs and use them with an NLB IngressController.
+- As a cluster administrator, I want to ensure EIPs are used with NLB on default router at install time.
+- As a cluster administrator, I want to reconfigure default router to use EIPs.
+- As a cluster administrator of OpenShift on AWS (ROSA), I want to use static IPs (and therefore AWS Elastic IPs) so that 
+  I can configure appropriate firewall rules.
+  I want the default AWS Load Balancer that they use (NLB) for their router to use these EIPs.
+- As a cluster administrator who has purchased a block of public IPv4 addresses, I want to use these IP addresses, so 
+  that I can avoid Amazon's new charge for public IP addresses.
+
+### Goals
+- Users are able to use EIP for a NLB Ingress Controller.
+- Users are able to create a NLB default Ingress Controller during install time with the EIPs specified during install.
+- The ELB of an existing IngressController should be able to be reconfigured to use specific EIPs by updating the IngressController CR and recreating the ELB with Load Balancer Type as `NLB`.
+
+### Non-Goals
+
+- Creation of EIPs in AWS.
```

Comment:
WRT the use case of `As a cluster administrator who has purchased a block of public IPv4 addresses, I want to use these IP addresses, so that I can avoid Amazon's new charge for public IP addresses.` :

The installer is designing to provide BYO IP support consuming Public IPv4 that the user brought to AWS by setting the Pool ID on [install-config.yaml `platform.aws.publicIpv4Pool`](https://github.com/openshift/installer/pull/7983/files#diff-7941d8da46384fc43b6eeae7d5c3319065b16940457822346c9ccb08cbcf47b4R109-R113), then, all resources created by the installer will consume Public IPs (EIP) from that pool.

To provide that automation, the terraform ([#7983](https://github.com/openshift/installer/pull/7983)) resources which claims EIP is setting that flag to consume from the pool when provided by the user. Furthermore, in the installer goal to migrate to CAPI (cluster-api), there is an implementation in progress to make the CAPA (cluster-api-provider-aws) to support the same capabilities performed on Terraform, which means, create the resources (NLB, Nat GW, instances) consuming from a custom pool, instead of delegating to the user to create the EIPs and associating it to the resources.
References on CAPA: [request](https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/4887) , [Example API](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/dad9f3ce8627b38000efe70ed3b0857c9d89590a/api/v1beta2/awsmachine_types.go#L184-L190) 

Is there any chance the cluster ingress operator (or some controller) to consume an Public IPv4 Pool ID available on installer, calculate the subnets, and creates (manage lifecycle) of the Elastic IPs required to create a service load balancer to the default router? So the user, and CI, wont need extra steps prior installing a cluster.

cc @jupierce @patrickdillon @r4f4

Reply:
@mtulio that looks outside the scope of this enhancement.

Reply:
>> Is there any chance the cluster ingress operator (or some controller) to consume an Public IPv4 Pool ID available on installer, calculate the subnets, and creates (manage lifecycle) of the Elastic IPs required to create a service load balancer to the default router? So the user, and CI, wont need extra steps prior installing a cluster.

No, this is not currently part of the epic. We may add it in the future if needed.